src/notification.rs

```diff
@@ -0,0 +1,481 @@
+    /// use serde::Deserialize;
+    ///
+    /// #[derive(Deserialize, Debug)]
+    /// struct Win32_ProcessStartTrace {
+    ///     ProcessID: u32,
+    ///     ProcessName: String,
+    /// }
+    ///
+    /// let mut stream = con.async_notification::<Win32_ProcessStartTrace>()?;
+    ///
+    /// let event = stream.next().await.unwrap()?;
+    /// #   Ok(()) // This query will fail when not run as admin
+    /// # }
+    /// ```
+    pub fn async_notification<T>(&self) -> WMIResult<impl Stream<Item = WMIResult<T>>>
+    where
+        T: serde::de::DeserializeOwned,
+    {
+        let mut query_text = build_query::<T>(None)?;
+        fetch_everything(&mut query_text);
+        self.async_raw_notification(query_text)
+    }
+
+    /// Subscribe to the T event, while filtering according to `filters`.
+    /// Returns a stream of WMIResult\<T\>.
+    ///
+    /// ```edition2018
+    /// # use wmi::*;
+    /// # use futures::{future::FutureExt, select};
+    /// # fn main() -> wmi::WMIResult<()> {
+    /// #   async_std::task::block_on(async {
+    /// #       select! { // End in 3 seconds or on event.
+    /// #           () = async_std::task::sleep(std::time::Duration::from_secs(3)).fuse() => Ok(()),
+    /// #           r = exec_async_query().fuse() => r
+    /// #       }
+    /// #   })
+    /// # }
+    /// #
+    /// # async fn exec_async_query() -> WMIResult<()> {
+    /// # use std::collections::HashMap;
+    /// # let con = WMIConnection::new(COMLibrary::new()?)?;
+    /// use futures::StreamExt;
+    /// use serde::Deserialize;
+    /// #[derive(Deserialize, Debug)]
+    /// struct __InstanceCreationEvent {
+    ///     TargetInstance: Win32_Process,
+    /// }
+    ///
+    /// #[derive(Deserialize, Debug)]
+    /// struct Win32_Process {
+    ///     ProcessID: u32,
+    /// }
+    ///
+    /// let mut filters = HashMap::new();
+    ///
+    /// filters.insert("".to_owned(), FilterValue::Within(1));
+    /// filters.insert("TargetInstance".to_owned(), FilterValue::IsA("Win32_Process"));
+    ///
+    /// let mut stream = con.async_filtered_notification::<__InstanceCreationEvent>(&filters)?;
+    ///
+    /// let event = stream.next().await.unwrap()?;
+    /// #   Ok(())
+    /// # }
+    /// ```
+    pub fn async_filtered_notification<T>(&self, filters: &HashMap<String, FilterValue>) -> WMIResult<impl Stream<Item = WMIResult<T>>>
+    where
+        T: serde::de::DeserializeOwned,
+    {
+        let mut query_text = build_query::<T>(Some(filters))?;
+        fetch_everything(&mut query_text);
+        self.async_raw_notification(query_text)
+    }
+}
+
+// This is a workaround to the following issue.
+// When querying for notifications, often times a query will contain object properties.
+// Since `wmi::de::meta:struct_name_and_fields` isn't able to get fields names from that inner object
+// we need to select the entire object in order to properly deserialize it.
+fn fetch_everything(query: &mut String) {
```

Comment:
I've added it, but I don't think this is the best solution.
I think that a query builder should be able to build **any** query, even if it is invalid.
But the current system isn't very good. Or at least it could be better.